### PR TITLE
runtime(-rs): resolve symlinked mount paths

### PR DESCRIPTION
**Pull request by manuelh-dev**
On some machines /var may be backed by a symlink, e.g., pointing to an NVMe storage medium.

When using emptyDir, the kubelet passes paths under /var/lib/kubelet to the shim. Looking up /proc/mounts finds their resolved NVMe paths, thus resulting in Error::NoMountEntry in get_linux_mount_info() in case that folder under /var is backed by a symlink.
With the literal path comparison then failing, memory-backed emptyDir mounts are misclassified as disk-backed mounts. Meaning: 'medium: Memory' is being ignored, and we silently fall back to, e.g., block-plain emptydir_mode.

This change canonicalizes the requested mount point before comparing it with /proc/mounts, and falls back to the original path if resolution fails. In case of a failure, we let our existing detection logic be the final judge to decide whether a matching mount exists or not.

### CODE REVIEW

**Comment by manuelh-dev** [comment]
hope the problem is clear, on a machine I was testing, /var is a symlink pointing to an NVMe, rootfs is small, and relevant paths where we need e.g. lots of storage are backed by large NVMEs. I was tinkering with emptyDirs and realized something went sideways.

**Comment by Apokleos** [comment]
> hope the problem is clear, on a machine I was testing, /var is a symlink pointing to an NVMe, rootfs is small, and relevant paths where we need e.g. lots of storage are backed by large NVMEs. I was tinkering with emptyDirs and realized something went sideways.
>
Thanks Manuel, yeah, Could you please provide a detail about the error message ?

**Comment by manuelh-dev** [comment]
> > hope the problem is clear, on a machine I was testing, /var is a symlink pointing to an NVMe, rootfs is small, and relevant paths where we need e.g. lots of storage are backed by large NVMEs. I was tinkering with emptyDirs and realized something went sideways.
> 
> Thanks Manuel, yeah, Could you please provide a detail about the error message ?

@Apokleos that's the thing, there is no visible error. In my cas,e `get_linux_mount_info()` returns `Error::NoMountEntry`, and `is_non_tmpfs_empty_dir()` intentionally consumes that result and returns true (see `Err(crate::mount::Error::NoMountEntry(_)) => true`). That leads to the volume being treated as disk-backed.

Before the fix, the runtime log shows creation and hot-plug of `.../uds-vol/disk.img`, and `/uds` being mapped from `/run/kata-containers/sandbox/storage/...` ...

After the fix, no block device is created for this volume. `/uds` is mapped from `/run/kata-containers/sandbox/ephemeral/uds-vol` and appears as tmpfs inside the container.

**Comment by manuelh-dev** [comment]
@Apokleos - it was kind of late last Friday when I opened this - in the meantime, looked into which other places in code may be affected by this pattern as well:
- `get_huge_page_option` in `src/runtime-rs/crates/resource/src/volume/hugepage.rs`. Maybe we could call our fixed `get_linux_mount_info` here.
- `IsTmpFSEmptyDir` in `src/runtime/virtcontainers/mount.go` calling `GetDevicePathAndFsTypeOptions` might face the same issue as the rust runtime -> this single seems to control the huge page behavior as well as the emptyDir behavior as well.

File: src/libs/kata-sys-util/src/mount.rs [hunk 1]
```diff
@@ -129,10 +130,21 @@ pub struct LinuxMountInfo {
...
 }
 
-/// Get the device and file system type of a mount point by parsing `/proc/mounts`.
+/// Get mount information for a mount point by parsing `/proc/mounts`.
 pub fn get_linux_mount_info(mount_point: &str) -> Result<LinuxMountInfo> {
+    // Mounting resolves symlinks in the target path, so `/proc/mounts` records
+    // the resolved path. In case `mount_point` contains a symlink, resolve it
+    // before comparing, or fall back to the original path if resolution fails.
+    // If resolution fails, compare the original path and let the scan determine
+    // whether a matching mount exists.
```
Comment (manuelh-dev) [inline src/libs/kata-sys-util/src/mount.rs:143]: same pattern as in go ... please let me know if you prefer a hard failure here.

**Comment by manuelh-dev** [comment]
> @Apokleos - it was kind of late last Friday when I opened this - in the meantime, looked into which other places in code may be affected by this pattern as well:
> 
> * `get_huge_page_option` in `src/runtime-rs/crates/resource/src/volume/hugepage.rs`. Maybe we could call our fixed `get_linux_mount_info` here.
> * `IsTmpFSEmptyDir` in `src/runtime/virtcontainers/mount.go` calling `GetDevicePathAndFsTypeOptions` might face the same issue as the rust runtime -> this single seems to control the huge page behavior as well as the emptyDir behavior as well.

Updated the PR - aligned the code path in runtime-rs as similar as possible to how it was already aligned in go (hugepage.rs get_huge_page_option now calls     match get_linux_mount_info(&mount_path) {), and also resolve symlinks on go side.

Please let me know what you think and if you see better ways  how to handle this.

On my system:
```
$ ls -lha /var/lib
lrwxrwxrwx 1 root root 22 Nov 20  2025 /var/lib -> /mnt/nvme1n1p1/var-lib
```

**Comment by Apokleos** [comment]
> > @Apokleos - it was kind of late last Friday when I opened this - in the meantime, looked into which other places in code may be affected by this pattern as well:
> > 
> > * `get_huge_page_option` in `src/runtime-rs/crates/resource/src/volume/hugepage.rs`. Maybe we could call our fixed `get_linux_mount_info` here.
> > * `IsTmpFSEmptyDir` in `src/runtime/virtcontainers/mount.go` calling `GetDevicePathAndFsTypeOptions` might face the same issue as the rust runtime -> this single seems to control the huge page behavior as well as the emptyDir behavior as well.
> 
> Updated the PR - aligned the code path in runtime-rs as similar as possible to how it was already aligned in go (hugepage.rs get_huge_page_option now calls match get_linux_mount_info(&mount_path) {), and also resolve symlinks on go side.
> 
> Please let me know what you think and if you see better ways how to handle this.
> 
> On my system:
> 
> ```
> $ ls -lha /var/lib
> lrwxrwxrwx 1 root root 22 Nov 20  2025 /var/lib -> /mnt/nvme1n1p1/var-lib
> ```
>
could you please post the /proc/mounts ? As it's hard for me directly construct this case locally.

**Comment by manuelh-dev** [comment]
> > > @Apokleos - it was kind of late last Friday when I opened this - in the meantime, looked into which other places in code may be affected by this pattern as well:
> > > 
> > > * `get_huge_page_option` in `src/runtime-rs/crates/resource/src/volume/hugepage.rs`. Maybe we could call our fixed `get_linux_mount_info` here.
> > > * `IsTmpFSEmptyDir` in `src/runtime/virtcontainers/mount.go` calling `GetDevicePathAndFsTypeOptions` might face the same issue as the rust runtime -> this single seems to control the huge page behavior as well as the emptyDir behavior as well.
> > 
> > 
> > Updated the PR - aligned the code path in runtime-rs as similar as possible to how it was already aligned in go (hugepage.rs get_huge_page_option now calls match get_linux_mount_info(&mount_path) {), and also resolve symlinks on go side.
> > Please let me know what you think and if you see better ways how to handle this.
> > On my system:
> > ```
> > $ ls -lha /var/lib
> > lrwxrwxrwx 1 root root 22 Nov 20  2025 /var/lib -> /mnt/nvme1n1p1/var-lib
> > ```
> 
> could you please post the /proc/mounts ? As it's hard for me directly construct this case locally.

Hope the following is a better illustration. Here is my symlink again:
```
$ls -ld /var/lib
lrwxrwxrwx 1 root root 22 Nov 20  2025 /var/lib -> /mnt/nvme1n1p1/var-lib

$readlink -f /var/lib
/mnt/nvme1n1p1/var-lib
```

Kubelet paths are normally passed around as `/var/lib/kubelet/...`, but `/proc/mounts` records the mounted tmpfs using the resolved path:
```
$ awk '$3 == "tmpfs" && $2 ~ "^/mnt/nvme1n1p1/var-lib/kubelet/pods/" { print; exit }' /proc/mounts
tmpfs /mnt/nvme1n1p1/var-lib/kubelet/pods/372ae739-2b7b-4b8d-88a6-7ebce78dba64/volumes/kubernetes.io~projected/kube-api-access-jm254 tmpfs rw,relatime,size=521243412k,inode64,noswap 0 0
```

The current (pre-fix) code effectively looks for the unresolved path:
```
$ awk '$2 == "/var/lib/kubelet/pods/372ae739-2b7b-4b8d-88a6-7ebce78dba64/volumes/kubernetes.io~projected/kube-api-access-jm254" { print }' /proc/mounts
# no output
```

But the same path resolves to the mount target that actually exists in `/proc/mounts`:
```
$ sudo readlink -f /var/lib/kubelet/pods/372ae739-2b7b-4b8d-88a6-7ebce78dba64/volumes/kubernetes.io~projected/kube-api-access-jm254
/mnt/nvme1n1p1/var-lib/kubelet/pods/372ae739-2b7b-4b8d-88a6-7ebce78dba64/volumes/kubernetes.io~projected/kube-api-access-jm254
```

**Comment by fidencio** [comment]
@coderabbit review

**Review by fidencio — APPROVED** [review]
lgtm, thanks @manuelh-dev!

JFYI, I've hit this on some of the internal systems that we can reserve at the employer.

**Comment by zvonkok** [comment]
Do we have a mock /proc/mounts to test all the corner cases?

File: src/libs/kata-sys-util/src/mount.rs [hunk 2]
```diff
@@ -148,11 +160,15 @@ pub fn get_linux_mount_info(mount_point: &str) -> Result<LinuxMountInfo> {
             ));
         }
 
-        if mount_point == fields[PROC_PATH_INDEX] {
+        if resolved_path == Path::new(fields[PROC_PATH_INDEX]) {
```
[the remark this answers is not in the document]
Comment (manuelh-dev) [reply src/libs/kata-sys-util/src/mount.rs]: ugh, good find - this is not a problem I am introducing, for my own little self-defense, but as this PR touches this area, we should fix this, see https://man7.org/linux/man-pages/man3/getmntent.3.html

**Comment by manuelh-dev** [comment]
> Do we have a mock /proc/mounts to test all the corner cases?

No - but we should. Using actual /proc/mount on all sorts of test hosts doesn't seem to be quite indeterministic. I will define an in-memory/in-test-code mock instead, this not only let's us cover relevant negative test cases, but also makes things more deterministic

**Comment by manuelh-dev** [comment]
@CodeRabbit review

**Review by danmihai1 — APPROVED** [review]
